File: Data Structures/Data Strucutes-2/Linked-List/linked-list.py
""" Linked list implementation with all the operations in python: """

class node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def insertAtBeginning(self, new_data):
        new_node = node(new_data)
        new_node.next = self.head
        self.head = new_node
        return

    #insert after a node:
    def insertAfter(self, previous_node, new_data):
        if previous_node is None:
            print("The given previous node must be in the Linked list!")
            return
        new_node = node(new_data)
        new_node.next = previous_node.next
        previous_node.next = new_node
        return
    
    #insert at the end:
    def insertAtEnd(self, new_data):
        new_node = node(new_data)
        #traverse to the end:
        if self.head is None:
            self.head = new_node
            return
        last = self.head
        while(last.next):
            last = last.next

        last.next = new_node
        return

    #delete node:
    def deleteNode(self, position):
        if self.head is None:
            return
        temp = self.head

        if position == 0:
            self.head = temp.next
            temp = None
            return
        #find the key to be deleted:
        for i in range(position - 1):
            temp = temp.next
            if temp is None:
                break
        #The key is not present:        
        if temp is None:
            return
        #Position already null
        if temp.next is None:
            return
        next = temp.next.next
        temp.next = None
        temp.next = next

    #search an element:
    def search(self, key):
        current = self.head
        while current is not None:
            if current.data == key:
                return True
            current = current.next
        return False

    #sort the Linked List:
    def sortLinkedList(self, head):
        current = head
        index = node(None)

        if head is None:
            return
        else:
            while current is not None:
                #index points to the node next to current:
                index = current.next
                while index is not None:
                    if current.data > index.data:
                        current.data, index.data = index.data, current.data
                    index = index.next
                current = current.next
    def printLinkedList(self):
        temp = self.head
        while(temp):
            print(str(temp.data), end=" ")
            temp = temp.next

if __name__ == "__main__":
    llist = LinkedList()
    llist.insertAtBeginning(2)
    llist.insertAtBeginning(4)
    llist.insertAtEnd(56)
    llist.printLinkedList()
    print(" ")
    llist.insertAfter(llist.head.next, 5)
    llist.printLinkedList()
    llist.deleteNode(3)
    print(" ")
    llist.printLinkedList()
    print(" ")
    if llist.search(4):
        print("Item found in the linked list")
    else:
        print("Item not found")

    #sorting the linked list:
    llist.sortLinkedList(llist.head)
    print(" ")
    llist.printLinkedList()

File: Data Structures/Data Strucutes-2/Linked-List/linked-list.cpp
/* Linked list operations in cpp: */

#include "iostream"

using namespace std;

struct Node{
    int data;
    struct Node * next = nullptr;
};

void insertAtBeginning(struct Node ** head_ref, int new_data){
    /* Allocate memory to a node:*/
    struct Node * new_node = (struct Node *)malloc(sizeof(struct Node));
    /*insert the data:*/
    new_node->data = new_data;
    new_node->next = (*head_ref);

    /* move head to new node:*/
    (*head_ref) = new_node;
}

/* insert a node after a node:*/
void insertAfter(struct Node *prev_node, int new_data){
    if(prev_node == nullptr){
        cout << "\n The previous node given cannot be null! \n";
        return;
    }

    struct Node * new_node = (struct Node *)malloc(sizeof(struct Node));
    new_node->data = new_data;
    new_node->next = prev_node->next;
    prev_node->next = new_node;
    return;
}

/*Insert at End:*/

void insertAtEnd(struct Node ** head_ref, int new_data){
    struct Node *new_node = (struct Node *)malloc(sizeof(struct Node));
    struct Node *last = *head_ref;

    new_node->data = new_data;
    new_node->next = nullptr;

    if (*head_ref == nullptr){
        *head_ref = new_node;
        return;
    }
    while(last->next != nullptr) last = last->next;

    last->next = new_node;
    return;

}

/* delete a certain node*/
void deleteNode(struct Node **head_ref, int key){
    struct Node *temp = *head_ref, *prev;
    
    /* The node is at the beginning: */
    if(temp != nullptr && temp->data == key){
        *head_ref = temp->next;
        free(temp);
        return;
    }

    /* find the key to be deleted:*/
    while(temp != nullptr && temp->data != key){
        prev = temp;
        temp = temp->next;
    }
    /* key is not present:*/
    if(temp == nullptr) return;

    /* remove the node:*/
    prev->next = temp->next;
    free(temp);
    return;
}

/* search a node:*/
bool searchNode(struct Node** head_ref, int key){
    struct Node *current = *head_ref;

    while(current != nullptr){
        if(current->data == key)return true;
        current = current->next;
    }
    return false;
}

/* sort the linked list with Bubble sort algorithm:*/
void sortLinkedList(struct Node ** head_ref){
    struct Node *current = *head_ref, *index = nullptr;
    
    int temp;

    if(head_ref == nullptr)return;
    else{
        while(current != nullptr){
            /* index points to the node next to current:*/
            index = current->next;
            while(index != nullptr){
                if(current->data > index->data){
                    temp = current->data;
                    current->data = index->data;
                    index->data = temp;
                }
                index = index->next;
            }
            current = current->next;

        }
    }
}
/* print the linked list:*/

void printLinkedList(struct Node * node){
    while(node != nullptr){
        cout << node->data << " ";
        node = node->next;
    }
}

int main(){
    struct Node * head = nullptr;
    insertAtEnd(&head, 1);
    insertAtBeginning(&head, 2);
    insertAtBeginning(&head, 3);
    insertAtEnd(&head, 4);
    insertAfter(head->next,5);
    
    cout << "\n Linked List ->\n";
    printLinkedList(head);
    cout <<"\nAfter deleting an element (3)-> \n";
    deleteNode(&head, 3);
    printLinkedList(head);
    if (searchNode(&head, 3)){
        cout<< "\n 3 is found\n";
    }else{
        cout <<"\n 3 is not found\n";
    }

    sortLinkedList(&head);
    cout <<"\nSorted Linked List -> \n";
    printLinkedList(head);

}